### DISCUSSION TITLE
Why are dimensions expected to be in a certain order?

### DISCUSSION THREAD

**Discussion by abarciauskas-bgse** [corteva/rioxarray#674] [Q&A]
👋🏽 I'm working on titiler-xarray with @sharkinsspatial and @vincentsarago and noticed that when trying reproject this zarr dataset s3://yuvipanda-test1/cmr/gpm3imergdl.zarr, we get an error (request to https://dev-titiler-xarray.delta-backend.com/tiles/0/0/0.png?reference=False&variable=precipitationCal&rescale=0,50&url=s3://yuvipanda-test1/cmr/gpm3imergdl.zarr):

```
{
  "detail": "Invalid dimension order. Expected order: ('y', 'x'). You can use `DataArray.transpose('y', 'x')` to reorder your dimensions. Data variable: precipitationCal"
}
```

We can simply transpose the coordinates, but we then have to absorb the cost of doing so on every request to our tiles endpoint. Is there a reason why the dimensions must be in the expected order?

<details>
<summary>Traceback:</summary>

```
  File "/Users/aimeebarciauskas/github/developmentseed/titiler-xarray/titiler/xarray/factory.py", line 249, in tiles_endpoint
    image = dst.tile(
  File "/Users/aimeebarciauskas/miniconda3/lib/python3.9/site-packages/rio_tiler/io/xarray.py", line 229, in tile
    ds = ds.rio.reproject(
  File "/Users/aimeebarciauskas/miniconda3/lib/python3.9/site-packages/rioxarray/raster_array.py", line 446, in reproject
    dst_data = self._create_dst_data(dst_height, dst_width)
  File "/Users/aimeebarciauskas/miniconda3/lib/python3.9/site-packages/rioxarray/raster_array.py", line 502, in _create_dst_data
    extra_dim = self._check_dimensions()
  File "/Users/aimeebarciauskas/miniconda3/lib/python3.9/site-packages/rioxarray/rioxarray.py", line 937, in _check_dimensions
    raise InvalidDimensionOrder(
rioxarray.exceptions.InvalidDimensionOrder: Invalid dimension order. Expected order: ('y', 'x'). You can use `DataArray.transpose('y', 'x')` to reorder your dimensions. Data variable: precipitationCal
```
</details>

**Comment by snowman2** [comment]
>  Is there a reason why the dimensions must be in the expected order?

That is the expected order of the array when calling `rasterio.warp.reproject`.

**Reply by abarciauskas-bgse** [reply]
Would it be reasonable to check that if the dimensions are out of order, to swap the order in rioxarray or rasterio? Or would this be brittle for some reason, like the name of the dimensions could be swapped (although this seems like fundamentally a data problem)?

**Reply by snowman2** [reply]
> Would it be reasonable to check that if the dimensions are out of order, to swap the order in rioxarray or rasterio?

The reasons why it isn't handled automatically:
- This could confuse the user if they get results that have the axis swapped on the output after re-projection.
- The input data is non-standard and it would be better to fix the input data.

**Reply by Kirill888** [reply]
> This could confuse the user if they get results that have the axis swapped on the output after re-projection.
The input data is non-standard and it would be better to fix the input data.

It should be possible to re-order bands post reprojection back to the original order, but this will result in extra cpu and peak memory consumption costs, and some code complexity costs too. Better to handle such transformation explicitly I feel. I think mandating `...,Y,X,...` order is reasonable, even demanding `...,Y,X` (`X` must be last dimension) is probably ok.
